Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+        return NSCalculationNoError;
+    }
+
+    if (IS_ZERO(rightOperand)) {
+        NSDecimalCopy(result, leftOperand);
+        return NSCalculationNoError;
+    }
+
+    NSDecimal left, right;
+
+    // Copy over decimal for structural manipulation.
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+
+    // normalize both of them
+    NSCalculationError error = NSDecimalNormalize(&left, &right, roundingMode);
+    if (error != NSCalculationNoError) {
+        return error;
+    }
+
+    // If they are both the same sign, then we can directly add the mantissa
+    if (IS_NEGATIVE(leftOperand) == IS_NEGATIVE(rightOperand)) {
+        // set the signs
+        NSDecimal* shortest;
+        NSDecimal* longest;
+        unsigned int shortLen;
+        unsigned int longestLen;
+        if (rightOperand->_length < leftOperand->_length) {
+            shortLen = rightOperand->_length;
+            shortest = &right;
+            longestLen = leftOperand->_length;
+            longest = &left;
+            NSDecimalCopy(result, &left);
+
+        } else {
+            shortLen = leftOperand->_length;
+            shortest = &left;
+            longestLen = rightOperand->_length;
+            longest = &right;
+            NSDecimalCopy(result, &right);
+        }
+
+        unsigned short finalResult[NSDecimalMaxSize + 1] = {};
+        unsigned int res = 0;
+
+        // Since the decimals are normalized, add the mantissa's and account for overflow.
+        for (int index = 0; index < shortLen; ++index) {
+            res += shortest->_mantissa[index] + longest->_mantissa[index];
+            finalResult[index] = res;
+            res = res >> 0x10;
+        }
+
+        // take care of the rest
+        for (int index = shortLen; index < longestLen; ++index) {
+            res += longest->_mantissa[index];
+            finalResult[index] = res;
+            res = res >> 0x10;
+        }
+
+        if (res != 0) {
+            if (longestLen >= NSDecimalMaxSize + 1) {
+                INVALIDATE_DECIMAL(result);
+                return NSCalculationOverflow;
+            }
+            finalResult[longestLen] = res;
+            result->_length = longestLen + 1;
+        }
+
+        // We need to fit
+        NSDecimalInternal internalResult = {};
+        NSDecimalCopy(&internalResult.decimal, result);
+        internalResult._mantissa = finalResult;
+        internalResult._remainder = 0;
+        internalResult._extendedLen = NSDecimalMaxSize + 1;
+
+        NSCalculationError error = _compactMantissa(&internalResult, roundingMode);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+
+        // set the result back
+        _InternalDecimalCopy(result, &internalResult);
+        NSDecimalCompact(result);
+        return NSCalculationNoError;
+    }
+
+    return _handleDifferenceOfSignAddition(result, &left, &right, roundingMode);
```

Comment:
nit: it's a little weird that one case is handled in a helper and the other isn't -shrug-